packages/alfa-dom/src/common/microsyntaxes.ts

```diff
@@ -0,0 +1,45 @@
+import { Map } from "@siteimprove/alfa-map";
+import { Parser } from "@siteimprove/alfa-parser";
+import { Err, Ok, Result } from "@siteimprove/alfa-result";
+
+/**
+ * @see https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#enumerated-attribute
+ */
+
+export enum EnumeratedValueError {
+  Missing = "missing",
+  Invalid = "invalid",
+}
+
+export function parseEnumeratedValue<RESULT>(
+  mapping: Map<string, RESULT>
+): Parser<string, RESULT, EnumeratedValueError> {
+  function parser(
+    str: string
+  ): Result<readonly [string, RESULT], EnumeratedValueError> {
+    const result = mapping.get(str.toLowerCase());
+
+    return str === ""
+      ? Err.of(EnumeratedValueError.Missing)
+      : result.isNone()
+      ? Err.of(EnumeratedValueError.Invalid)
+      : Ok.of(["", result.get()] as const);
+  }
+
+  return parser;
+}
```

Comment:
This would be well-suited as `Attribute#enumerate(): Option<string>`. The current API is a little unfortunate in that it requires a very specific type of mapping (specifically a `Map<K, V>`) to be given up front, which isn't the only conceivable thing we would want to do with an enumerated attribute. In general, we should aim to design APIs that make as few assumptions as possible in order to allow callers the most flexibility.

With the above API, the specific use case you have could then be supported by `attribute.enumerate().flatMap(keyword => mapping.get(keyword))` or even this:

```ts
attribute.enumerate().flatMap(keyword => {
  switch (keyword) {
    case "foo":
    case "bar":
      return Some.of(keyword);
    default:
      return None;
  }
});
```

`#enumerate()` would handle all the repetitiveness of lowercasing the attribute value and returning `None` when it's empty; the caller then knows that they receive a keyword that is ready to map to a state or do whatever else the caller wants to do.